drivers/syslog/syslog_stream.c

```diff
@@ -0,0 +1,303 @@
+ *   buflen     - The number of bytes in the buffer.
+ *
+ * Returned Value:
+ *   On success, the number of characters written is returned. A negated
+ *   errno value is returned on any failure.
+ *
+ ****************************************************************************/
+
+static ssize_t syslog_stream_write(FAR struct syslog_channel_s *channel,
+                                   FAR const char *buffer, size_t buflen)
+{
+  FAR struct syslog_stream_s *chan =
+    (FAR struct syslog_stream_s *)channel;
+  FAR struct lib_outstream_s *stream =
+    (FAR struct lib_outstream_s *)chan->stream;
+  ssize_t nwritten;
+  irqstate_t flags;
+
+  flags = enter_critical_section();
+  nwritten = stream->puts(stream, buffer, buflen);
+  leave_critical_section(flags);
+  return nwritten;
+}
+
+/****************************************************************************
+ * Name: syslog_stream_putc
+ *
+ * Description:
+ *   This is the low-level, single character, system logging interface
+ *   provided for the driver interface.
+ *
+ * Input Parameters:
+ *   channel    - Handle to syslog channel to be used.
+ *   ch         - The character to add to the SYSLOG (must be positive).
+ *
+ * Returned Value:
+ *   On success, the character is echoed back to the caller. A negated errno
+ *   value is returned on any failure.
+ *
+ ****************************************************************************/
+
+static int syslog_stream_putc(FAR struct syslog_channel_s *channel, int ch)
+{
+  FAR struct syslog_stream_s *chan =
+    (FAR struct syslog_stream_s *)channel;
+  FAR struct lib_outstream_s *stream =
+    (FAR struct lib_outstream_s *)chan->stream;
+  ssize_t nbytes;
+  uint8_t uch;
+  irqstate_t flags;
+
+  flags = enter_critical_section();
+  uch = (uint8_t)ch;
+  nbytes = stream->puts(stream, &uch, 1);
+  leave_critical_section(flags);
+
+  /* Check if the write was successful.  If not, nbytes will be
+   * a negated errno value.
+   */
+
+  if (nbytes < 0)
+    {
+      return (int)nbytes;
+    }
+
+  return ch;
+}
+
+/****************************************************************************
+ * Name: syslog_stream_force
+ *
+ * Description:
+ *   Force output in interrupt context.
+ *
+ * Input Parameters:
+ *   channel    - Handle to syslog channel to be used.
+ *   ch         - The character to add to the SYSLOG (must be positive).
+ *
+ * Returned Value:
+ *   On success, the character is echoed back to the caller.  A negated
+ *   errno value is returned on any failure.
+ *
+ ****************************************************************************/
+
+static int syslog_stream_force(FAR struct syslog_channel_s *channel, int ch)
+{
+  FAR struct syslog_stream_s *chan =
+    (FAR struct syslog_stream_s *)channel;
+  FAR struct lib_outstream_s *stream =
+    (FAR struct lib_outstream_s *)chan->stream;
+  ssize_t nbytes;
+  uint8_t uch;
+
+  uch = (uint8_t)ch;
+  nbytes = stream->puts(stream, &uch, 1);
```

Comment:
```suggestion
  lib_stream_put(chan->stream, ch);
```

Reply:
Current blkoutstrea and mtdoutstream don't have a put interface.
Should we add put interface for them?

Reply:
really? I think it's better to improve mtd/blk stream to support char version.

Reply:
Okay, I will add putchar interface for both streams.

Reply:
https://github.com/apache/nuttx/pull/7715, see here